energymeters/integrationpluginenergymeters.cpp

```diff
@@ -0,0 +1,394 @@
+    m_connectionStateTypeIds.insert(sdm630ThingClassId, sdm630ConnectedStateTypeId);
+
+    m_voltageL1StateTypeIds.insert(pro380ThingClassId, pro380VoltageL1StateTypeId);
+    m_voltageL1StateTypeIds.insert(sdm630ThingClassId, sdm630VoltageL1StateTypeId);
+
+    m_voltageL2StateTypeIds.insert(pro380ThingClassId, pro380VoltageL2StateTypeId);
+    m_voltageL2StateTypeIds.insert(sdm630ThingClassId, sdm630VoltageL2StateTypeId);
+
+    m_voltageL3StateTypeIds.insert(pro380ThingClassId, pro380VoltageL3StateTypeId);
+    m_voltageL3StateTypeIds.insert(sdm630ThingClassId, sdm630VoltageL3StateTypeId);
+
+    m_currentL1StateTypeIds.insert(pro380ThingClassId, pro380CurrentL1StateTypeId);
+    m_currentL1StateTypeIds.insert(sdm630ThingClassId, sdm630CurrentL1StateTypeId);
+
+    m_currentL2StateTypeIds.insert(pro380ThingClassId, pro380CurrentL2StateTypeId);
+    m_currentL2StateTypeIds.insert(sdm630ThingClassId, sdm630CurrentL2StateTypeId);
+
+    m_currentL3StateTypeIds.insert(pro380ThingClassId, pro380CurrentL3StateTypeId);
+    m_currentL3StateTypeIds.insert(sdm630ThingClassId, sdm630CurrentL3StateTypeId);
+
+    m_activePowerStateTypeIds.insert(pro380ThingClassId, pro380CurrentPowerEventTypeId);
+    m_activePowerStateTypeIds.insert(sdm630ThingClassId, sdm630CurrentPowerStateTypeId);
+
+    m_frequencyStateTypeIds.insert(pro380ThingClassId, pro380FrequencyStateTypeId);
+    m_frequencyStateTypeIds.insert(sdm630ThingClassId, sdm630FrequencyStateTypeId);
+
+    m_powerFactorStateTypeIds.insert(pro380ThingClassId, pro380PowerFactorStateTypeId);
+    m_powerFactorStateTypeIds.insert(sdm630ThingClassId, sdm630PowerFactorStateTypeId);
+
+    m_totalEnergyConsumedStateTypeIds.insert(pro380ThingClassId, pro380TotalEnergyConsumedEventTypeId);
+    m_totalEnergyConsumedStateTypeIds.insert(sdm630ThingClassId, sdm630TotalEnergyConsumedEventTypeId);
+
+    m_totalEnergyProducedStateTypeIds.insert(pro380ThingClassId, pro380TotalEnergyProducedStateTypeId);
+    m_totalEnergyProducedStateTypeIds.insert(sdm630ThingClassId, sdm630TotalEnergyProducedStateTypeId);
+
+    m_discoverySlaveAddressParamTypeIds.insert(pro380ThingClassId, pro380DiscoverySlaveAddressParamTypeId);
+    m_discoverySlaveAddressParamTypeIds.insert(sdm630ThingClassId, sdm630DiscoverySlaveAddressParamTypeId);
+
+    m_registerMaps.insert(pro380ThingClassId, pro380RegisterMap);
+    m_registerMaps.insert(sdm630ThingClassId, sdm630RegisterMap);
+
+    // FIXME leads to crash, probably an issue in the modbus rtu resource
+    /*connect(hardwareManager()->modbusRtuResource(), &ModbusRtuHardwareResource::modbusRtuMasterRemoved, this, [=] (const QUuid &modbusUuid){
+        qCDebug(dcEnergyMeters()) << "Modbus RTU master has been removed" << modbusUuid.toString();
+
+        // Check if there is any device using this resource
+        foreach (EnergyMeter * meter, m_energyMeters) {
+            if (meter->modbusRtuMasterUuid() == modbusUuid) {
+                Thing *thing = m_energyMeters.key(meter);
```

Comment:
Please make sure you will set the modbusRtuMaster within the Meter to nullptr and make sure it will not be accessed within the meter class without a nullptr check, otherwise you end up in a situation where you want to update the registers, the resource has been removed and you access the modbusRtuMaster.

Reply:
done